lib/asciidoctor/js/opal_ext/reader.rb

```diff
@@ -0,0 +1,209 @@
+                  else
+                    logger.warn message_with_context %(unexpected end tag '#{this_tag}' at line #{inc_lineno} of include #{target_type}: #{inc_path}), :source_location => cursor, :include_location => include_cursor
+                  end
+                end
+              elsif inc_tags.key?(this_tag = $2)
+                tags_used << this_tag
+                # QUESTION should we prevent tag from being selected when enclosing tag is excluded?
+                tag_stack << [(active_tag = this_tag), (select = inc_tags[this_tag]), inc_lineno]
+              elsif !wildcard.nil?
+                select = active_tag && !select ? false : wildcard
+                tag_stack << [(active_tag = this_tag), select, inc_lineno]
+              end
+            elsif select
+              # NOTE record the line where we started selecting
+              inc_offset ||= inc_lineno
+              inc_lines << l
+            end
+          end
+        rescue
+          logger.error message_with_context %(include #{target_type} not readable: #{inc_path}), :source_location => cursor
+          return replace_next_line %(Unresolved directive in #{@path} - include::#{expanded_target}[#{attrlist}])
+        end
+        unless tag_stack.empty?
+          tag_stack.each do |tag_name, _, tag_lineno|
+            logger.warn message_with_context %(detected unclosed tag '#{tag_name}' starting at line #{tag_lineno} of include #{target_type}: #{inc_path}), :source_location => cursor, :include_location => (create_include_cursor inc_path, expanded_target, tag_lineno)
+          end
+        end
+        unless (missing_tags = inc_tags.keys.to_a - tags_used.to_a).empty?
+          logger.warn message_with_context %(tag#{missing_tags.size > 1 ? 's' : ''} '#{missing_tags.join ', '}' not found in include #{target_type}: #{inc_path}), :source_location => cursor
+        end
+        shift
+        if inc_offset
+          parsed_attrs['partial-option'] = true unless base_select && wildcard && inc_tags.empty?
+          # FIXME not accounting for skipped lines in reader line numbering
+          push_include inc_lines, inc_path, relpath, inc_offset, parsed_attrs
+        end
+      else
+        begin
+          # NOTE read content first so that we only advance cursor if IO operation succeeds
+          inc_content = read_include_content inc_path, target_type
+          shift
+          push_include inc_content, inc_path, relpath, 1, parsed_attrs
+        rescue
+          logger.error message_with_context %(include #{target_type} not readable: #{inc_path}), :source_location => cursor
+          return replace_next_line %(Unresolved directive in #{@path} - include::#{expanded_target}[#{attrlist}])
+        end
+      end
+      true
+    end
+  end
+
+  # If a VFS is defined, Asciidoctor will use it to resolve the include target.
+  # Otherwise use the file system or the network to read the file.
+  def read_include_content inc_path, target_type
```

Comment:
@mojavelinux `preprocess_include_directive` method is almost identical, I've just replaced `File.open` with the new `read_include_content` method.